keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,394 @@
+├── framework.yaml
+├── params.yaml
+└── common
+    └── common.yaml
+└── templates
+    ├── backup.yaml
+    ├── clear-data.yaml
+    ├── deployment.yaml
+    ├── init.yaml
+    ├── load-data.yaml
+    ├── pvc.yaml
+    ├── restore.yaml
+    └── service.yaml
+```
+
+#### framework.yaml
+
+`framework.yaml` is the base definition of an framework. It follows the following format, extracted from the MySQL example:
+
+```yaml
+name: "my-framework"
+description: A test example framework.
+version: "5.7"
+kudoVersion: ">= 0.2.0"
+kubeVersion: ">= 1.14"
+maintainers:
+  - Bob <bob@example.com>
+  - Alice <alice@example.com>
+url: https://github.com/myframework/myframework
+tasks:
+  deploy:
+    resources:
+      - pvc.yaml
+      - deployment.yaml
+      - service.yaml
+      - deployment2.yaml
+      - job.yaml
+    patches:
+      - deploy-patch.yaml
+    patchesStrategicMerge:
+      - super-weird-deploy-patch.yaml
+  init:
+    resources:
+      - init.yaml
+  pv:
+    resources:
+      - pvc.yaml
+  backup:
+    resources:
+      - init.yaml
+    patches:
+      - backup.yaml
+  restore:
+    resources:
+      - init.yaml
+    patches:
+      - restore.yaml
+  load-data:
+    resources:
+      - init.yaml
+    patches:
+      - load-data.yaml
+  clear-data:
+    resources:
+      - init.yaml
+    patches:
+      - clear-data.yaml
+  query:
+    resources:
+      - job.yaml
+plans:
+  deploy:
+    steps:
+      - name: deploy
+        tasks:
+          - deploy
+      - name: init
+        tasks:
+          - init
+      - name: cleanup
+        tasks:
+          - init
+        delete: true
+  backup:
+    steps:
+      - name: pv
+        tasks:
+          - pv
+      - name: backup
+        tasks:
+          - backup
+      - name: cleanup
+        tasks:
+          - backup
+        delete: true
+  restore:
+    steps:
+      - name: restore
+        tasks:
+          - restore
+      - name: cleanup
+        tasks:
+          - restore
+        delete: true
+```
+
+While subsequent sections go into deeper detail, the top level keys of the framework are:
+
+- **version**: String defining the version of a given framework
+- **tasks**: A map of tasks that can be run. These are the atomic runnable unit of a KUDO framework, and are made up of a series of YAML manifests. These are defined more in detail in [Tasks](#tasks).
+- **plans**: A map of plans that can be run. These are the core partitioning unit of a KUDO framework. A plan is intended to run a single "operations task" for an framework, such as backup, restore, deployment, or update. This is defined in detail in [Plans](#plans)
+
+This file undergoes a Go template pass on Instance instantiation before being parsed. This is described more in detail in [Extensions and Bases](#extensions-and-bases)
+
+#### params.yaml
+
+The `params.yaml` file is a struct that defines parameters for framework. This can articulate descriptions, defaults, and triggers etc., In the MySQL example, this looks like:
```

Comment:
I might miss some context but having a `params.yaml` file tells me that I will find parameters with values and descriptions. I was surprised when I saw that a param also has a field `triggers`.

Reply:
Ah, yeah, need to describe triggers. I'll add that in!